store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java

```diff
@@ -339,11 +347,22 @@ public void deliver(LmtpEnvelope env, InputStream in, int sizeHint) throws Unrec
         Blob blobEEW = null;
         // duplicating input stream for EEW
         try {
-            final String content = IOUtils.toString(in); // pop content from original input stream
-            IOUtils.closeQuietly(in);
-            in = IOUtils.toInputStream(content); // push content back to input stream
-            final String contentEEW = ExternalEmailWarning.getInstance().getUpdatedContent(content);
-            inEEW = IOUtils.toInputStream(contentEEW);
+            for (int i = 0; i < CHARSET_NAMES.length; i++) {
+                final String charsetName = CHARSET_NAMES[i];
+                // copy input stream and evaluate content encoding
+                final byte[] ba = IOUtils.toByteArray(in);
+                IOUtils.closeQuietly(in);
+                in = new ByteArrayInputStream(ba);
+                final String content = new String(ba, charsetName);
+                // if current encoding is suitable, use it; otherwise, continue or use default
+                if (content.contains(GARBLED_CHARACTER) && i < CHARSET_NAMES.length - 1) {
```

Comment:
Why do we need this "i < CHARSET_NAMES.length - 1" check here?

Reply:
As per the understanding of logic, for the last element of array i.e. `MimeConstants.P_CHARSET_DEFAULT`, Code has to be always executed in else block, that is the reason `i < CHARSET_NAMES.length - 1` is implemented.

Reply:
We may write with single if block .
`if (!content.contains(GARBLED_CHARACTER) || i == CHARSET_NAMES.length - 1) {
}`

Reply:
@zimsuchitgupta @shrutig0510 modified as suggested